pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -0,0 +1,1150 @@
+		Annotations: pod.Annotations,
+	}
+
+	dnsServers, dnsSearches, err := m.runtimeHelper.GetClusterDNS(pod)
+	if err != nil {
+		return nil, err
+	}
+	podSandboxConfig.DnsOptions = &runtimeApi.DNSOption{
+		Servers:  dnsServers,
+		Searches: dnsSearches,
+	}
+
+	hostname, _, err := m.runtimeHelper.GeneratePodHostNameAndDomain(pod)
+	if err != nil {
+		return nil, err
+	}
+	podSandboxConfig.Hostname = &hostname
+
+	if pod.Spec.SecurityContext != nil {
+		securityContext := pod.Spec.SecurityContext
+		podSandboxConfig.Linux = &runtimeApi.LinuxPodSandboxConfig{
+			NamespaceOptions: &runtimeApi.NamespaceOption{
+				HostNetwork: &securityContext.HostNetwork,
+				HostIpc:     &securityContext.HostIPC,
+				HostPid:     &securityContext.HostPID,
+			},
+		}
+	}
+
+	cgroupParent := ""
+	podResourceLimits := make(map[api.ResourceName]int64)
+	podResourceRequests := make(map[api.ResourceName]int64)
+	portMappings := []*runtimeApi.PortMapping{}
+	for _, c := range pod.Spec.Containers {
+		for name, limit := range c.Resources.Limits {
+			if l, ok := podResourceLimits[name]; ok {
+				podResourceLimits[name] = l + limit.MilliValue()
+			} else {
+				podResourceLimits[name] = limit.MilliValue()
+			}
+		}
+		for name, req := range c.Resources.Requests {
+			if l, ok := podResourceRequests[name]; ok {
+				podResourceRequests[name] = l + req.MilliValue()
+			} else {
+				podResourceRequests[name] = req.MilliValue()
+			}
+		}
+
+		opts, err := m.runtimeHelper.GenerateRunContainerOptions(pod, &c, "")
+		if err != nil {
+			return nil, err
+		}
+
+		for _, port := range opts.PortMappings {
+			hostPort := int32(port.HostPort)
+			containerPort := int32(port.ContainerPort)
+			protocol := toRuntimeProtocol(port.Protocol)
+			portMappings = append(portMappings, &runtimeApi.PortMapping{
+				HostIp:        &port.HostIP,
+				HostPort:      &hostPort,
+				ContainerPort: &containerPort,
+				Protocol:      &protocol,
+				Name:          &port.Name,
+			})
+		}
+
+		cgroupParent = opts.CgroupParent
+	}
+
+	if cgroupParent != "" {
+		if podSandboxConfig.Linux != nil {
+			podSandboxConfig.Linux.CgroupParent = &cgroupParent
+		} else {
+			podSandboxConfig.Linux = &runtimeApi.LinuxPodSandboxConfig{
+				CgroupParent: &cgroupParent,
+			}
+		}
+	}
+
+	if len(portMappings) > 0 {
+		podSandboxConfig.PortMappings = portMappings
+	}
+
+	if len(podResourceLimits) > 0 || len(podResourceRequests) > 0 {
+		podSandboxConfig.Resources = &runtimeApi.PodSandboxResources{}
+		var podCPULimit, podMemoryLimit, podCPURequest, podMemoryRequest int64
+		for k, v := range podResourceLimits {
+			switch k {
+			case api.ResourceCPU:
+				podCPULimit = v
+			case api.ResourceMemory:
+				podMemoryLimit = v
+			}
+		}
+		for k, v := range podResourceRequests {
+			switch k {
+			case api.ResourceCPU:
+				podCPURequest = v
+			case api.ResourceMemory:
+				podMemoryRequest = v
+			}
+		}
+		if podCPULimit != 0 || podCPURequest != 0 {
+			cpuResource := &runtimeApi.ResourceRequirements{}
```

Comment:
What if we just initialize with &runtimeApi.ResourceRequirements{}, and apply values despite whether it is 0 or not, which seems cleaner to me.